### ISSUE TITLE
Malformed encoding found in chunked-encoding

### ISSUE THREAD

**Issue by dnivras** [curl/curl#3490]
I use curl to make REST API calls to our server.

Expect no error in the call.

[curl -V output]
curl 7.59.0 (x86_64-redhat-linux-gnu) libcurl/7.59.0 OpenSSL/1.1.0h zlib/1.2.11 libpsl/0.20.1 (+libidn2/2.0.4) libssh/0.7.5/openssl/zlib nghttp2/1.31.1
Release-Date: 2018-03-14
Protocols: dict file ftp ftps gopher http https imap imaps ldap ldaps pop3 pop3s rtsp scp sftp smb smbs smtp smtps telnet tftp 
Features: AsynchDNS IPv6 Largefile GSS-API Kerberos SPNEGO NTLM NTLM_WB SSL libz TLS-SRP HTTP2 UnixSockets HTTPS-proxy PSL Metalink

### operating system
Linux localhost.localdomain 4.16.3-301.fc28.x86_64 #1 SMP Mon Apr 23 21:59:58 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
Fedora release 28 (Twenty Eight)

[curl.txt](https://github.com/curl/curl/files/2781838/curl.txt)
Debug dump showing that the server sends the correct encoding length of 0x68 followed by 0x0d 0x0a followed by the chunk of length 0x68 followed again by 0x0d 0x0a.  According to the Wiki article on chunked encoding (https://en.wikipedia.org/wiki/Chunked_transfer_encoding) this is a valid response from the server.

~~~
<= Recv data, 0000000110 bytes (0x0000006e)
0000: 36 38 0d 0a 7b 22 61 75 74 68 54 6f 6b 65 6e 22 68..{"authToken"
0010: 3a 22 33 35 33 33 34 34 62 65 2d 61 31 64 35 2d :"353344be-a1d5-
0020: 34 35 30 37 2d 61 39 66 62 2d 64 33 32 34 30 31 4507-a9fb-d32401
0030: 32 34 38 30 63 33 22 2c 22 61 75 74 68 54 6f 6b 2480c3","authTok
0040: 65 6e 45 78 70 69 72 79 22 3a 22 31 35 36 32 39 enExpiry":"15629
0050: 36 30 39 35 36 34 32 32 22 2c 22 74 77 6f 46 41 60956422","twoFA
0060: 54 6f 6b 65 6e 22 3a 6e 75 6c 6c 7d 0d 0a       Token":null}..
CommAPICurl.cc:notifyResponse:267: Copying response to buffer (LMS Req id: 0 offset: 0 nbytes: 104):
== Info: Malformed encoding found in chunked-encoding
== Info: stopped the pause stream!
== Info: Closing connection 0

**Comment by dnivras** [comment]
Can I send you a linux binary?  It is a public URL so you should be able to run it.

Can you replay a tcpdump stream?  How do I record it?

Thanks,
Arvind

**Comment by dnivras** [comment]
[curl.txt](https://github.com/curl/curl/files/2786151/curl.txt)
The whole debug dump:

**Comment by jzakrzewski** [comment]
I may be wrong but isn't the chunked encoding supposed to end with an empty chunk (like `30 0d 0a 0d 0a`)? I don't see that in the dump.

**Comment by dnivras** [comment]
Thanks!  I believe that's it - when I run it from the command line curl, it gets those 5 bytes and doesn't complain and leaves the connection intact.  And when I run my test program it doesn't get the empty chunk.
Though why the server behaves like this I don't get at all - it does seem to be influenced by some setting from the client side.

**Comment by bagder** [comment]
Unfortunately this then isn't a curl bug but is a broken server, which might be harder to actually address. You really should report this issue to the server admins of that host. Maybe it will help by pointing them to this issue.

You might have to resort to doing a work-around where you simply ignore this specific error when working with this server. The difference in responses between curl and libcurl requests are highly curious. Is the outgoing requests perhaps different in some way that makes the server react differently?

Side-note: The chunked-encoding parser in Firefox silently ignores if there's no final zero chunk arriving and just pretends everything is still fine. I struggled hard once during a period to tighten its parser strictness for that, but had to give up since there were a few fairly popular server-side tools that generated this kind of broken chunked data and thus my attempt caused problems for too may innocent Firefox users and I had to revert it. I **really** do not want to take libcurl into that "quagmire".
